Regent version

LonelyCat124 opened the pull request:
Figured now that the Regent version is working I'd create the WIP request for discussion.

Next on my action plan is to work out how to time the code, and try to optimise various things, and then move on to mappers etc.

At current, I doubt there is much parallelism available for the code, I'm not sure if the vectorisation is present or anything.

Which versions should I compare to for performance, and any preference on which machines I should benchmark performance on?

arporter commented:
I'd go for perfomance results on the best machines that would allow us/you to write a paper. I would suggest the best Intel socket you can run on (SKL on SFP?) and compare with the OpenMP version. We could also compare with the MPI version but that is 'fresh out of the box' and may have performance issues.
If Regent lets you do GPU then I would also go for Glados.
If you or @sergisiso was interested, we could probably benchmark the OpenCL version too (on either CPU or GPU) but this would require a bit of work because we haven't done any optimisation on that yet.
Given that the CPU results in our last paper were done on my desktop, we can always run on there in order to compare with the results from the paper. That will give us some continuity.

sergisiso commented:
As Andy suggested it would be good to start with node-wide comparisons with the OpenMP version, so you have a performance baseline and results to compare to. Timing are done just for the time-stepping part (no initialization) but if you ask to output data every x iterations, these are done inside the timer region, so I would compare without producing output files for now. Any multi-core system should be OK at least for development.

When the ECP project starts then we will have again some EuroEXA time to spare which can be used to port NemoLite2D to OmpSs. A comparison among both task-based programming models could also be interesting if this is something you are still considering doing.

LonelyCat124 commented:
Hi Both,

Sounds good - I'm currently working on building Regent/Legion on scafellpike without the distributed memory support, and without CUDA support for now (it was failing with CUDA as I think it needs an older version of clang). I'll compare to the OpenMP version for now then, and then once its optimised reasonably well for a single node, go from there. I assume there'll be a reasonably steep learning curve for me to get maximal performance out but I have no idea yet. I'll probably keep results on this PR just so everything is is one place
Have a good weekend!

LonelyCat124 commented:
Ok, after a period of dividing by 10e5 instead of 10e6, some initial results:
git branch: db2efde477eede8c9ef7525f71841695fab0fd3a
namelist: 3000 steps, write after 6000 (no I/O), 258x258
Interactive node of Scafell pike
```
OpenMP version (1 thread): 39.134s, 38.499s, 38.986s, average: 38.873s
OpenMP version (32 threads): 3.2999s, 3.3751s, 3.6403s, average: 3.4384s
Regent version (no optimisation, limited parallelism, legion master as of 30/03, build options limited as its a pain to build on scafellpike, took me ): 33.973406s,  33.940966s, 33.909764s, average: 33.941379s
```

I'll run some differently sized tests after coffee, but faster than serial is the current success story...

LonelyCat124 commented:
Note mostly for me, but discovering a few things as I go along:
legion/language/regent.py algorithm.rg -lg:prof 1 -ll:cpu 32 -lg:prof_logfile prof.gz

This runs the legion profiler. I believe `ll:cpu` controls the number of CPUs used to run the code, so here are my investigations how changing that flag affects the code performance. Tilesize chosen was 128, single run, no repeats, default namelist:
```
ll:cpu X    |    Runtime
1           |     39.627414
2           |     26.402976
4           |     22,754922
8           |     23.482493
16          |     24.085219
32          |     27.410079
```
Performance is behaving roughly as I'd expect at this point, limited parallelism available (and large swathes of the code being serial) means we get almost 2x performance before gradually losing performance again

LonelyCat124 commented:
So looking at the profile of the small tile sizes it shows a near-serial execution, with significant amount of time being spent in the runtime vs the tasks.
![image](https://user-images.githubusercontent.com/3043914/78157294-b6602900-7437-11ea-8314-6138356e6e32.png)

 Task runtime for this is shown to be on the order of 307us for the update_velocity tasks (2nd and 3rd loops) while the utilities associated with this task are taking about 2500us:
![image](https://user-images.githubusercontent.com/3043914/78157836-69c91d80-7438-11ea-9032-72552950ce08.png)

I assume I can provide more information to the runtime/clean up various operations to improve this performance and allow parallelism to be applied.

rupertford commented:
> So looking at the profile of the small tile sizes it shows a near-serial execution, with significant amount of time being spent in the runtime vs the tasks.
> ![image](https://user-images.githubusercontent.com/3043914/78157294-b6602900-7437-11ea-8314-6138356e6e32.png)
> 
> Task runtime for this is shown to be on the order of 307us for the update_velocity tasks (2nd and 3rd loops) while the utilities associated with this task are taking about 2500us:
> ![image](https://user-images.githubusercontent.com/3043914/78157836-69c91d80-7438-11ea-9032-72552950ce08.png)
> 
> I assume I can provide more information to the runtime/clean up various operations to improve this performance and allow parallelism to be applied.

It may be slow, but the graphs look pretty :-)

LonelyCat124 commented:
I tried a couple things this morning, which seem relatively successful:
1) I removed all the partitions which are unused, in case they helped the runtime make decisions faster (I don't believe this made any difference)
2) I used the `-ll:util X` flag to enable the runtime to run utility threads in parallel as well as just the computation threads.
Note that these results are on a 32 core node with hyperthreading turned on, and only a single run unless it took many attempts to avoid crashes, so an accuracy of +/- 5% is probably reasonable.
Tilesize 64 (max parallelism is 16):

| cpu | util | runtime (s) |
|-----|------|-------------|
| 1   | 1    | 128.106171  |
| 1  | 2 | 
| 2   | 1    | 89.085581   |
| 2   | 2    | 52.798897   |
| 2   | 4    | 53.760885   |
| 2   | 8    | crash       |
| 4   | 2    | 41.288844   |
| 4   | 4    | 24.390105   |
| 8   | 2    | 35.490550   |
| 8   | 4    | 19.717295   |
| 8   | 8    | 14.499734   |
| 16  | 4    | 19.505642   |
| 16  | 8    | 13.525895   |
| 16  | 16   | 14.383243   |
| 32  | 8    | 15.341041   |
| 32  | 12  | 13.725529   |
| 32 | 16 | 14.257430 |
| 32 | 32 | 16.636243 |

Tilesize 32 (max parallelism is 64), most combinations i tried for ulti number threads were not successful:
| cpu | util | runtime (s) |
|-----|------|-------------|
| 1   | 2    | 2776.646604 |
| 16 | 4 | 164.715427 |
| 32 | 4 | 156.570639 |
| 32 | 8 | 99.486980 | 
| 32 | 16 | 77.258642 |
| 32 | 32 | 98.226028 |

This tilesize is too small to get performance regardless of util threads.

I'll edit this with profiles from a step of some of the faster runs with a tilesize of 64, and some other tilesize options as well. If they show high CPU utilisation, then the next step is to add parallelism to other sections of the code, and aim for high CPU utilisation throughout. There are limiting issue with vectorisation right now for sure, as the code explicitly says it doesn't vectorise some sections of the code

LonelyCat124 commented:
So the new version is built, without any of the additional parallelism I was running with the old code. From discussions I've had, the recommended task granularity is ~1ms, which for the 258x258 case may cause some difficulties due to the OpenMP running at ~1.13ms/step. For the 1024x1024 case its much easier, as there is about 10ms/timestep in total, so most of that should be comparable if the code is written well.

Some numbers and data for the new version on the 258x258 testcase:
`update_velocity_ufield` and `update_velocity_vfield` tasks take around 3ms, the only other main task being executed with significant runtime is `update_velocity_and_t_height`, with a runtime of about 0.6ms, and the other `update...` and `calculate_sea_surface_height_t` tasks with a runtime of 0.2-0.3ms. The other tasks are all smaller and <= 0.1ms.

I have removed the code that checks if the `model_write` task should be executed every step, and instead only do any work in the task if the model needs writing - the launching and check takes around 24us, and can apparently be done in parallel to the u and v height updates.

Once I have a consistent sequence of task launching, and don't create any new partitions, I can use the `__demand(__trace)` option to tell the runtime that this sequence of tasks will be the same every time, which reduces overhead. With trace, the run gives this profile (3 CPU 8 util):
![image](https://user-images.githubusercontent.com/3043914/78358744-994c6700-75ab-11ea-9087-c6587033a688.png)

For this run setup, the trace makes relatively little difference, but I expect with more tasks I'll see more difference:
![image](https://user-images.githubusercontent.com/3043914/78359338-d06f4800-75ac-11ea-8e23-d0c953630492.png)


Plan next is to try adding parallelism back into the `update_velocity` tasks, and see how that performs on this and larger testcases with various tilesizes (256, 128, 84) at various scales and profile those.

LonelyCat124 commented:
Ok, some results and profiling data for the newest version. This version (a4cb47568e4866cc4b66cb910c85759f8ac241fc) has some parallelism again, and uses `__trace` for the loops as we have the same loops every iteration.
258x258, tilesize 84:
| cpu | util | runtime (s) |
|-----|------|-------------|
| 1 |  1 | 37.213844 | 
| 8 | 4 | 10.699351 |
| 16 | 4 | 9.135276 |
| 16 | 8 | ** 9.117819** |
| 24 | 4 | 9.221184 |
| 28 | 4 | 9.836077 |
| 32 | 4 | 10.232317 |


1024x1024, tilesize 256 (probably too big but fine for now). Only the velocity field calculations are parallel at the moment:
| cpu | util | runtime (s) |
|-----|------|-------------|
| 1 | 1 | 454.279888 |
| 8 | 4 | 111.755526 |
| 8 | 8 | 112.166264 |
| 16 | 4 | 81.939163 |
| 16 | 8 | 83.096003 |
| 24 | 4 | 72.215316 |
| 28 | 4 | **71.131787** |
| 28 | 8 | 72.819294 |
| 32 | 4 | 80.140603|

1024x1024, tilesize 128, variation is +/- ~1s:
| cpu | util | runtime (s) |
|-----|------|-------------|
| 1 | 1 | 464.896514 |
| 8 | 4 |  119.411185 |
| 16 | 4 | 83.532849 |
| 24 | 4 | **72.376847** |
| 28 | 4 | 78.554573 |
| 28 | 8 | 74.752962 |
| 32 | 4 | 72.943586  |
| 32 | 8 | 76.609244 |

For 258x258 the regent versoin is 2.6x slower than the OpenMP version and for 1024x1024 2.04x slower. Since the code is likely not vectorising with regent right now, and most of the code is still serial I think this is a positive start. For smaller testcases than 258x258 I'm skeptical of the performance that can be gained with regent however.

The profile (1024x1024 with 32 worker and 8 util threads, the 10 steps shows 2830ms for the full application for later comparison) shows that perhaps the timing interferes somewhat with the tracing and potential progression, as the first timestep is significantly slower than the following onces, and the util trace shows nothing happens for the entirety of the `model_write` call:
![image](https://user-images.githubusercontent.com/3043914/78564229-07936280-7814-11ea-9d32-9c193caeb6ad.png)

My expectation is that this is due to the call to `__fence` before the timing call and the main timestepping loop prevents the inital loop tracing occuring while we're doing I/O, and this results in the first timestep taking significantly longer (you can see the utilisation and duration of the first timestep is significantly longer): 
![image](https://user-images.githubusercontent.com/3043914/78564919-0a428780-7815-11ea-89ac-76f547bd64d8.png)

My plan is to test removing the timing code for profiling purposes, there is also an argument to be made that `model_write` can be done in parallel to computation to some point, though without data duplication (which I think we can try later) it can only be part of the first step (and this is all currently blocked by the fence.

I think the other potential target for tasking parallelism is `update_velocity_and_t_height` at this problem size. I think other performance options are available for other tasks, with `__demand(__vectorize)` and `__demand(__openmp)` options available for certain loops.

LonelyCat124 commented:
Actually still not totally convinced by the accuracy of the timing right now, I'd expect it to be out by a few hundred ms right now - For a 10 step run it reports 0.45s and the profiler reports 0.31s for the computational region - I'm going to attempt to use a task to create the timings.

LonelyCat124 commented:
So the last day or so I did the stages from before, and did some breakdowns of how the various changes affect performance (in terms of CPU seconds) of various tasks.

Relative performance is equal to `single task runtime / this runtime` (so sort of speedup)

Firstly, the single task results (all were run with `ll:cpu 24 ll:util 8 ll:ht_sharing 0` unless otherwise specified). This means the tilesizes were set to == the domain sizes, and openmp was disabled:
| Task | Sum runtime (ms) | Relative performance | Task standard deviation | Mean task time |
| ---------- | -------| ----- | ----- | ----- |
| calculate_sea_surface_t | 7349 | 1 | 0 | 7349 |
| update_velocity_v_field | 52347 | 1 | 0 | 52347 |
| update_velocity_u_field | 50912 | 1 | 0 | 50912 |
| update_velocity_and_t_height | 7025 | 1 | 0 | 7025 |
| update_v_height | 4682 | 1 | 0 | 4682 |
| update_u_height | 3614 | 1 | 0 | 3614 |


Dividing only the velocity field calculations to a task size of 256:
| Task | Sum runtime (ms) | Relative performance | Task standard deviation | Mean task time |
| ---------- | -------| ----- | ----- | ----- |
| calculate_sea_surface_t | 8251 | 0.89 | 0 | 8251 |
| update_velocity_v_field | 68306 | 0.77 | 255.4892 | 4269.125 |
| update_velocity_u_field | 87592 | 0.58 | 951.964 | 5474.5 |
| update_velocity_and_t_height | 7149 | 0.98 | 0 | 7149 |
| update_v_height | 3726 | 1.3 | 0 | 3726 |
| update_u_height | 3729 | 0.97 | 0 | 3729 |

Also tiling the sea_surface to task sizes of 512:
| Task | Sum runtime (ms) | Relative performance | Task standard deviation | Mean task time |
| ---------- | -------| ----- | ----- | ----- |
| calculate_sea_surface_t | 12357| 0.59 | 91.9 | 3089 |
| update_velocity_v_field | 80139 | 0.65| 1256.663 | 5008.688 |
| update_velocity_u_field | 95088 | 0.54 | 1546.382 | 5943 |
| update_velocity_and_t_height | 7553 | 0.93 | 0 | 7553|
| update_v_height | 4124 | 1.14 | 0 | 4124 |
| update_u_height | 3807 | 0.94 | 0 | 3807 |

Disabling the `ll:ht_sharing 0` option makes a significant performance degradation it seems like on the interactive nodes - I'm not going to include those results here as they're relatively unimportant at this time but they are recorded in my local profile 9.

Adding `ll:ocpu 2 ll:othr 4` (2 OpenMP cores and 4 OpenMP threads per "core"):
| Task | Sum runtime (ms) | Relative performance | Task standard deviation | Mean task time |
| ---------- | -------| ----- | ----- | ----- |
| calculate_sea_surface_t | 12336 | 0.60 | 118.716 | 3084 |
| update_velocity_v_field | 76656  | 0.68 | 820.7028 | 4791 |
| update_velocity_u_field | 78245 | 0.65| 930.5927 | 4890 |
| update_velocity_and_t_height | 7369 | 0.95 | 0 | 7369 |
| update_v_height | 5208 | 0.90 | 0 | 5208 |
| update_u_height | 3912 | 0.92 | 0 | 3912 |

Any variation of ~15% could be random/noise i think, so only worth considering larger changes for most of this.
So the tiled tasks seem to have a parallel efficiency of around 55-65% (ignoring any other runtime overheads, this is purely the time spent computing the tasks) which seems pretty inefficient, so it seems clear there is some optimisation to be done. I assume some of the performance loss might be data movement/data layout related, but it seems quite high. I'm going to try to see if there's a better way to obtain this data than my current manual inspection of the profiles, so I can automate it and gather the data more quickly (and thus reduce the noise). It is however worth noting that this is parallelising each of those workloads of 16 tasks, so still an 8 time performance gain in real terms (ignoring load imbalance).
For the OpenMP tasks, assuming the reported values are correct, there actually seems relatively little benefit to parallelising in this way, despite the performance results suggesting improved performance, so I'm somewhat skeptical of the new data on that at this point, but the suggestion may be alternative methods of parallelism (tasks enabling data parallelism) may be better.

I think at this point I'm going to dig in more depth to the examples provided with regent, as I feel like I understand enough to be able to go through their examples and hopefully adapt things from their stencil codes to improve performance of this example

LonelyCat124 commented:
So I had a look into the regent example, and it doesn't provide much more insight with how to get performance for our real-world case (as it doesn't worry about boundaries like we have to with the tmask).
We could (during setup) create extra variables, such as `tmask_plus_west_tmask`, which is 1 if they sum to >0 or 0 otherwise inside the grid, and then create a partition using `partition(grid.tmask_plus_west_tmask, ispace(int1d,2))`, and then only call the `update_u_height` function on that index space. I'm not clear at this time whether that would actually be worth any performance however (potentially it may allow regent to ok vectorising these functions, but it might be able to anyway).

I did change the parallelisation of the update functions from OpenMP to partitioning + tasks, and it looks like this improves performance again, 1024x1024, 256 tilesize1, 512 tilesize2, no openmp no `ll:ht_sharing` flag
| cpu | util | runtime (s) |
|-----|------|-------------|
| 8 | 4 | 87.008315 |
| 8 | 8 | 88.410507 |
| 16 | 4 | 55.966071 |
| 16 | 8 | 57.488733 |
| 24 | 4 | **48.591909** |
| 24 | 8 | 51.080303 |
| 28 | 4 | 49.079133 |
| 32 | 4 | 63.129037 |
| 32 | 8 | 62.554999 |


I'm going to look how to tile as many of the remaining untiled tasks as possible to maximise the parallelism available - this is mostly an issue for the `bc_flather` tasks as they do not tile well necessarily (or at least, they have dependencies Northwards or Eastwards to other `bc_flather` tasks). I also tried briefly testing using full columns/rows as "tiles" (i.e. partitioning the domain with x=N*N, y = 1 or vice versa) but this didn't  show improvement over the tiles - it might be mapper dependent which I've not yet looked at.

LonelyCat124 commented:
So I fully partitioned and ran the task code, but I need a `  __fence(__execution, __block)` between the `update_*vel_boundary`  and the `bc_flather` functions. I'm not entirely clear why this dependency isn't detected correctly, I'll look next week, I think because the second task does read and write it isn't correctly detected.

LonelyCat124 commented:
So I think this must be an error of understanding on my part. The dependency is shown in the dependency profiles between the `update_*vel_boundary` functions and the `bc_flather*` functions.. I could consider joining them into the same task later but I don't think that helps with this issue at all. I can't see why it should result in correctness errors so I'll have to analyse the code manually and check if there's any other potentially missing depdendencies,

LonelyCat124 commented:
Ok, so my investigations this morning seems to suggest if i have the barrier anywhere between the `update_velocity_*field` calls and the `bc_flather*` call, then at 12000 timesteps the answer is still correct.

The weirdness I can't work out is that there are dependencies from all of the `update_velocity_*field` calls to a `update_*vel_boundary`, and from the `update_*vel_boundary` to the `bc_flather*` call.
Now the tiling in this case (258x258, tilesize1 is 84, tilesize2 is 128) works out that the `update_velocity_*field` calls are a 3x3 index space and the other calls are on a 2x1 or 1x2 index space.

In principle, looking at solely the `update_velocity_ufield` functions, we should have:
6 tasks with 1 out dependency to a `update_uvel_boundary` and 3 with 2 out dependencies to `update_uvel_boundary` tasks, as there is overlap between the middle of that 3x3 grid and the two columns (or rows, whichever way round it is).
From looking at the profiler output with spy, i have the expected dependencies between tasks, so I'm unclear as to what is going wrong, I guess I'll have to make an issue on the main legion github to see if I'm wrong in my understanding or not.

LonelyCat124 commented:
Regent guys have pointed me to the bounds checking tool, and that appears to be what I'm messing up, seem to have got my (i,j) indexing and N/M values the wrong way round in some functions, so I'm going to work at fixing that today.

LonelyCat124 commented:
Ok, big update today as I fixed a bunch of broken code due to incorrect bounds and me confusing array index ordering, and added a way to add config to choose the two tilesizes (called t1 and t2), and enabling me to automate that. New version was correct for the standard 258x258 case @ 12k steps. New results for the 1024x1024 testcase for 3000 steps:
| cpu | util | t1 | t2 | runtime (s) |
|-----|------|-------------|
| 8 | 4 | 128 | 256 | 132.713828 |
| 8 | 4 | 128 | 512 | 84.102708 |
| 8 | 4 | 256 | 256 | 79.776436 |
| 8 | 4 | 256 | 512 | 83.633248 |
| 8 | 4 | 512 | 256 | 78.291087 |
| 8 | 4 | 512 | 512 | 80.921360 |
| 8 | 8 | 128 | 256 | 106.041390 |
| 8 | 8 | 128 | 512 | 84.328823 |
| 8 | 8 | 256 | 256 | 80.059917 |
| 8 | 8 | 256 | 512 | 85.370750 |
| 8 | 8 | 512 | 256 | 78.507205 |
| 8 | 8 | 512 | 512 | 84.182151 |
| 16 | 4 | 128 | 256 | 137.465456 |
| 16 | 4 | 128 | 512 | 61.419971 |
| 16 | 4 | 256 | 256 | 56.881541 |
| 16 | 4 | 256 | 512 | 51.827261 |
| 16 | 4 | 512 | 256 | 75.953292 |
| 16 | 4 | 512 | 512 | 101.888134 |
| 16 | 8 | 128 | 256 | 80.927667 |
| 16 | 8 | 128 | 512 | 54.034564 |
| 16 | 8 | 256 | 256 | 48.853561 |
| 16 | 8 | 256 | 512 | 53.908241 |
| 16 | 8 | 512 | 256 | 76.691590 |
| 16 | 8 | 512 | 512 | 80.450142 |
| 24 | 4 | 128 | 256 | 134.704545 |
| 24 | 4 | 128 | 512 | 63.192689 |
| 24 | 4 | 256 | 256 | 53.119201 |
| 24 | 4 | 256 | 512 | **44.676072** |
| 24 | 4 | 512 | 256 | 78.081146 |
| 24 | 4 | 512 | 512 | 80.550091 |
| 24 | 8 | 128 | 256 | 111.617629 |
| 24 | 8 | 128 | 512 | 56.771746 |
| 24 | 8 | 256 | 256 | 48.801747 |
| 24 | 8 | 256 | 512 | 45.703661 |
| 24 | 8 | 512 | 256 | 78.713841 |
| 24 | 8 | 512 | 512 | 81.016810 |
| 32 | 4 | 128 | 256 | 136.805519 |
| 32 | 4 | 128 | 512 | 63.425933 |
| 32 | 4 | 256 | 256 | 60.989809 |
| 32 | 4 | 256 | 512 | 57.893273 |
| 32 | 4 | 512 | 256 | 88.517404 |
| 32 | 4 | 512 | 512 | 86.807379 |
| 32 | 8 | 128 | 256 | 81.438982 |
| 32 | 8 | 128 | 512 | 49.432265 |
| 32 | 8 | 256 | 256 | 56.939823 |
| 32 | 8 | 256 | 512 | 56.572439 |
| 32 | 8 | 512 | 256 | 98.377323 |
| 32 | 8 | 512 | 512 | 88.444890 |

Consistently t1=256, t2=512 seems to be best, I forgot to run with 28 & 4 as thread counts, that runtime is 44.503357. Trying 2 & 4 with t1=256, t2=1024 I get 71.437556s, so partitioning that to get more parallelism is useful.

The target runtime is : 38.123s for the openMP version so its getting pretty close now!

LonelyCat124 commented:
Update performance runs:

`2048x2048, OpenMP result: 173.34s`

Regent version, 258x258 problem size:
| cpu | util | t1 | t2 | runtime (s) |
|-----|------|----|----|-------------|
| 8 | 4 | 64 | 128 | 13.404898 |
| 8 | 4 | 64 | 256 | 10.554796 |
| 8 | 4 | 84 | 128 | 10.526245 |
| 8 | 4 | 84 | 256 | 11.280113 |
| 8 | 4 | 128 | 128 | 9.317589 |
| 8 | 4 | 128 | 256 | 11.707944 |
| 8 | 8 | 64 | 128 | 9.527381 |
| 8 | 8 | 64 | 256 | 10.882884 |
| 8 | 8 | 84 | 128 | 9.319556 |
| 8 | 8 | 84 | 256 | 11.263736 |
| 8 | 8 | 128 | 128 | 9.452076 |
| 8 | 8 | 128 | 256 | 11.424574 |
| 16 | 4 | 64 | 128 | 14.182195 |
| 16 | 4 | 64 | 256 | 9.240148 |
| 16 | 4 | 84 | 128 | 10.928087 |
| 16 | 4 | 84 | 256 | 9.474686 |
| 16 | 4 | 128 | 128 | 9.328603 |
| 16 | 4 | 128 | 256 | 10.838518 |
| 16 | 8 | 64 | 128 | 9.706025 |
| 16 | 8 | 64 | 256 | 8.961431 |
| **16** | **8** | **84** | **128** | **8.151848** |
| 16 | 8 | 84 | 256 | 9.727640 |
| 16 | 8 | 128 | 128 | 9.300655 |
| 16 | 8 | 128 | 256 | 10.751734 |
| 24 | 4 | 64 | 128 | 15.693305 |
| 24 | 4 | 64 | 256 | 9.945138 |
| 24 | 4 | 84 | 128 | 12.004421 |
| 24 | 4 | 84 | 256 | 8.871253 |
| 24 | 4 | 128 | 128 | 10.812963 |
| 24 | 4 | 128 | 256 | 11.223481 |
| 24 | 8 | 64 | 128 | 12.918279 |
| 24 | 8 | 64 | 256 | 8.951064 |
| 24 | 8 | 84 | 128 | 10.110386 |
| 24 | 8 | 84 | 256 | 9.277569 |
| 24 | 8 | 128 | 128 | 10.016608 |
| 24 | 8 | 128 | 256 | 11.248323 |
| 32 | 4 | 64 | 128 | 15.953315 |
| 32 | 4 | 64 | 256 | 10.916144 |
| 32 | 4 | 84 | 128 | 11.989465 |
| 32 | 4 | 84 | 256 | 10.372351 |
| 32 | 4 | 128 | 128 | 11.551597 |
| 32 | 4 | 128 | 256 | 11.956823 |
| 32 | 8 | 64 | 128 | 11.551495 |
| 32 | 8 | 64 | 256 | 10.757710 |
| 32 | 8 | 84 | 128 | 10.306915 |
| 32 | 8 | 84 | 256 | 10.707490 |
| 32 | 8 | 128 | 128 | 12.028848 |
| 32 | 8 | 128 | 256 | 12.290834 |

Will add the 2048x2048 results later once they've run

LonelyCat124 commented:
**Skylake results thread **
OpenMP version compiled on the Skylakes with GCC 8.1 . Regent built with llvm 3.8 and gcc 8.1. 3k timesteps, no I/O

| Size | R1 | R2 | R3 | Avg |
|----|---|---|---|---|
| 2048 x 2048 | 135.15s | 135.00s | 136.47s | **135.54s** |
| 1024 x 1024 | 34.850s | 34.863s | 34.650s | **34.788s** |
| 258 x 258 | 2.9081s | 2.9355s | 3.1678s | **3.0038s** |

LonelyCat124 commented:
Need this hash d7b19022d03de1cf388228b578712689279b1e8

LonelyCat124 commented:
Ok - so I created a new version which uses partitions to start removing some if statements. The issue with this right now is that this disables the ability to use index_launches, as these are constructed via a `cross_product`.
This is noted in regent issue: https://github.com/StanfordLegion/legion/issues/845

The performance is therefore worse (which I believe mostly to be due to that, but I'm not sure right now). Here are the current results for 2048x2048 (t1 256 t2 512):
| cpu | util | runtime (s) |
|-----|------|-------------|
| 24 | 4 | 174.488797 |
| 24 | 8 | 171.114973 |
| 28 | 4 | 221.849474 |
| 28 | 8 | 193.408288 |
| 32 | 4 | 182.750033 |
| 32 | 8 | 169.754451 |

For reference, the prior version had best 157.54s, and OpenMP had 135.54s. I have a few more if statements to remove (and actually will result in separating some functions) but hopefully these changes combined with Regent upstream changes could result in at worst, similar performance but more in-fitting with the model I think.

One big issue I still am seeing, is that the biggest hotspots still can't vectorise, due to branches such as :
```
      var sign_v_n : double = 0.0
      if(v_n >= 0.0) then
        sign_v_n = 1.0
      else
        sign_v_n = -1.0
      end
```

Weirdly, the intel compiler vectorises this entire loop, even on AVX2, so there must be alternatives. @sergisiso do you know a vectorisable equivalent to `SIGN` in C/C++?

Once I solve that, there's still a few more if statements that are point dependent which would be nice to avoid which I need to think about. I could create a "mask" for each of the if statements manually but that seems a bit gross, but it would work.

One issue right now is the overheads from creating these partitions is non-zero, but I think I can delete them once they're done with potentially, I've not tested that yet.

arporter commented:
> Weirdly, the intel compiler vectorises this entire loop, even on AVX2, so there must be alternatives.
> @sergisiso do you know a vectorisable equivalent to SIGN in C/C++?

If only you were using Fortran :-) (Sorry, couldn't resist.)

LonelyCat124 commented:
Apparently an option is:
```
template <typename T> int sgn(T val) {
    return (T(0) < val) - (val < T(0));
}
```
I think its not quite the same as fortran SIGN as I think that doesn't return 0 if the input is 0? Wheras this would return 0. Thats not the same right @arporter ?

I also realised i ran with the wrong t2 size for this case, rerunning it now.

arporter commented:
> 
> 
> Apparently an option is:
> 
> ```
> template <typename T> int sgn(T val) {
>     return (T(0) < val) - (val < T(0));
> }
> ```
> 
> I think its not quite the same as fortran SIGN as I think that doesn't return 0 if the input is 0? Wheras this would return 0. Thats not the same right @arporter ?
> 
Erm, I don't understand what that code does! Fortran SIGN returns the ABS of its first argument with the sign of the second argument. Does that help?

LonelyCat124 commented:
Ah, I was using it differently because I couldn't do it in one operation.

The original code is:
```
--        uu_w = (0.5_go_wp - SIGN(0.5_go_wp, u_w)) * un%data(ji,jj)              + &
--             & (0.5_go_wp + SIGN(0.5_go_wp, u_w)) * un%data(ji-1,jj)
```
so we get either +0.5 or -0.5. I think my question is what happens if u_w is 0? I assume based on my previous code:
```
      var sign_u_w : double = 0.0
      if(u_w >= 0.0) then
        sign_u_w = 1.0
      else
        sign_u_w = -1.0
      end
      var uu_w = (0.5 - 0.5 * sign_u_w) * velocity_full[point].u_now
               + (0.5 + 0.5 * sign_u_w) * velocity_full[point + {-1,0}].u_now
```
that if `u_w == 0` then `SIGN` returns 0.5.

So probably the equivalent C code to sign (for doubles lets say) is:
```
double sign( double val, double signed){
  double sign = (double) ((0.0 <= signed) - (signed < 0.0))
  return sign * val;
}
```
Since if signed is >= 0.0, then we get `1-0` for sign, and if `signed < 0.0` we get `0-1`, then multiply that by value. Does that seem more sensible @arporter @sergisiso ? I can try to create this in terra also. This relies on messy C that a true statement gives 1 and a false statement gives 0.

So for the 2048 x 2048 case the actual results (t1 256 t2 1024) are:
| cpu | util | runtime (s) |
|-----|------|-------------|
| 24 | 4 | 157.904989 |
| 24 | 8 | 161.569791 |
| 28 | 4 | 148.074085 |
| 28 | 8 | 147.562335 |
| 32 | 4 | 161.376745 |
| 32 | 8 | 170.386401 |

So the performance is actually ~10s better than before, and closer to the OpenMP (gcc version)!

arporter commented:
Yes, that seems sensible to me.

LonelyCat124 commented:
It needs to be `abs(val) * signing` but yeah. So i need to find vectorisable abs next :D

LonelyCat124 commented:
This works:
```
__demand(__inline)
task sign( value: double, signed : double) : double
  var signing : double = [double] ( [int64](signed >= 0.0) - [int64](0.0 > signed) )
  var abs_value = rfabs(value)
  return abs_value * signing
end
```

I want to benchmark its performance vs if statements tomorrow.

sergisiso commented:
@AidanChalk Seems you already found a good alternative. Anyway, in the C version of the kernel we use `copysing` from <math.h> https://en.cppreference.com/w/c/numeric/math/copysign but I have not checked if it gets properly vectorized by the compilers.

LonelyCat124 commented:
Ah ok! I'll if regentlib has that already included (since that has vectorisable versions) and if so thats much easier.

LonelyCat124 commented:
There is copysign doh!

LonelyCat124 commented:
Once i test with copysign, the remaining blockers to vectorisation in these loops are things like:
```
    IF(un%grid%tmask(ji,jj-1) <=0 .OR. un%grid%tmask(ji+1,jj-1) <= 0) THEN
       dudy_s = 0.0_go_wp !slip boundary
    ELSE
       dudy_s = (un%data(ji,jj) - un%data(ji,jj-1)) / (un%grid%dy_u(ji,jj) + un%grid%dy_u(ji,jj-1)) * &
            & (hu%data(ji,jj) + sshn_u%data(ji,jj) + hu%data(ji,jj-1) + sshn_u%data(ji,jj-1))
    END IF
```

This is already reversed to 
```
      if( grid_region[point + {0,-1}].tmask > int1d(0)
          and grid_region[point + {1,-1}].tmask > int1d(0)) then
```
but essentially this wants to be a mask where I multiple by 0.0 or 1.0 depending on the condition being true.
I can perhaps do `var mask : double = [double] ( grid_region[point + {0,-1}].tmask > int1d(0)          and grid_region[point + {1,-1}].tmask > int1d(0))` but that feels a bit hacky.
This check is essentially `grid_region[point + {0,-1}].tmask + grid_region[point + {1,-1}].tmask == int2d(2)`, with possible values of `{-2, -1, 0, 1, 2}`, so I could do:
```
var mask : double = [double]( ([int64](grid_region[point + {0,-1}].tmask + grid_region[point + {1,-1}].tmask + int1d(1)) /3 )
```

Unfortunately this is 2 additions, 1 integer division (YUCK) and 2 casts (and integer cast to double is likely expensive, while int1d to int64 might be compiler optimised, unclear), wheras the more ugly:
```
var mask : double = [double](grid_region[point + {0,-1}].tmask + grid_region[point + {1,-1}].tmask == int1d(2))
```
Looks more performant (no division, 1 cast).

I don't really like any of these solutions though, but the only other option would be to split this loop up more which sounds even worse.

LonelyCat124 commented:
Ok, so I changed to use the `copysign` function, and the code loses ~10s on the large testcase again.
I'm going to try a standalone benchmark for this of if statements vs copysign with/without vectorisation to explain this.

LonelyCat124 commented:
So I craeted a test code, with a warmup loop (as without that it was heavily biased):
```
import "regent"


copysign = regentlib.copysign(double)

local c = regentlib.c
local format = require("std/format")

fspace point{
  a : double
}

task get_time() : double

  return c.legion_get_current_time_in_micros()
end

task main()

 var reg = region( ispace(int1d, 250000000), point)
 var reg2 = region( ispace(int1d, 250000000), point)

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  var start_time = get_time()
  __fence(__execution, __block)

 __forbid(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  __fence(__execution, __block)
  var finish_time = get_time()
  __fence(__execution, __block)
  var time_taken = finish_time - start_time
  format.println("non-vectorize copysign: {}", double(time_taken)/1000000.0)
  __fence(__execution, __block)
 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  start_time = get_time()
  __fence(__execution, __block)

 __demand(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  __fence(__execution, __block)
  finish_time = get_time()
  __fence(__execution, __block)
  time_taken = finish_time - start_time
  format.println("vectorize copysign: {}", double(time_taken)/1000000.0)
  __fence(__execution, __block)

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  start_time = get_time()
  __fence(__execution, __block)

  for point in reg do
    var z : double = 0.0
    if reg2[point].a >= 0.0 then
      z = 1.0
    else
      z = -1.0
    end
    reg[point].a = z * reg[point].a
  end

  __fence(__execution, __block)
  finish_time = get_time()
  __fence(__execution, __block)
  time_taken = finish_time - start_time
  format.println("if statement: {}", double(time_taken)/1000000.0)
  __fence(__execution, __block)

end
regentlib.start(main)
```

The resulting performance here:
```
non-vectorize copysign: 1.374571
vectorize copysign: 1.373119
if statement: 1.278637
```
This suggests the performance of `copysign` is worse than just if statements in regent. @sergisiso have you done a similar comparison for C/C++? If not I can put one together and see if this mirrors the C/C++ performance.

LonelyCat124 commented:
I made a C version:
```
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

void fill(double *reg_a, double *reg_b){

  for( int i = 0; i < 25000000; i++){
    reg_a[i] = 0.75;
    reg_b[i] = -1.0;
  }

}

int main(){


  double *reg_a, *reg_b;

  reg_a = malloc(sizeof(double) * 25000000);
  reg_b = malloc(sizeof(double) * 25000000);

  fill(reg_a, reg_b);
  for( int i = 0; i < 25000000; i++){
    reg_a[i] = copysign(reg_a[i], reg_b[i]);
  }
  fill(reg_a, reg_b);
  double start_time = omp_get_wtime();
  for( int i = 0; i < 25000000; i++){
    reg_a[i] = copysign(reg_a[i], reg_b[i]);
  }
  double finish_time = omp_get_wtime();
  printf("copysign took %f\n", (finish_time - start_time));

  fill(reg_a, reg_b);
  start_time = omp_get_wtime();
  for( int i = 0; i < 25000000; i++){
    double z = 0.0;
    if(reg_b[i] >= 0.0) z = 1.0;
    else z = -1.0;
    reg_a[i] = z * reg_a[i];
  }
  finish_time = omp_get_wtime();
  printf("if statements took %f\n", (finish_time - start_time));

  free(reg_a);
  free(reg_b);

}
```

gcc9 `gcc test_sign.c -fopenmp -O3` :
```
copysign took 0.031722
if statements took 0.033240
```

intel19, `icc test_sign.c -qopenmp -std=c99 -O3`:
```
copysign took 0.035331
if statements took 0.033093
```

Performance is pretty similar between copysign and using if statements - is this an unfair comparison? Its MUCH slower with regent either way, which could be mapper related but I'm not sure.

sergisiso commented:
@AidanChalk I never compared it but what you did looks sensible, and I suspect that beneath the copysing there is a similar expression than the if statements so it make sense they have the same performance.

Regarding Regent I don't fully understand the how it works yet, is there some jit overhead for the lua instructions? or has this already been lowered to machine code when the whole task main is created? Also C probably sees the definition of copysing in the maths header and inlines it, is the regentlib capable of doing this or does it add a function call or some indirection to a block of code or something?

LonelyCat124 commented:
It should have been lowered to machine code, I think it should inline it as well but I'm not sure. I think copysign should not use if statements (instead using the logic i showed yesterday or something similar based on the sign bit) in math library but i'm not going to go check :p.

I expect that something in the default mapper means the data layout is somehow bad? Or the loop is bad? I'm not clear though - I will check with developers perhaps.

LonelyCat124 commented:
I started looking at the mappers stuff, and decided that to really do that properly I needed to learn about the runtime properly, so I'm going through  https://www.youtube.com/playlist?list=PLUNK9XcztK7xjXfppL9hIpVv2ukp7A4tG as quickly as is sensible. I have about 14 pages of notes so far and I feel I understand (admittedly 5 year old version of) the runtime better, but memory layout I'm still unclear on - I think its handled by the mappers (and not the low level runtime) but I'm not 100% yet, I think once I get to the sections on Regions and the Realm Events & Reservations videos i should know properly.

LonelyCat124 commented:
The regent developers have now explained to me whats going wrong with the copysign instruction to be so slow at the end of https://github.com/StanfordLegion/legion/issues/848 . There's a follow up issue on this here: https://github.com/StanfordLegion/legion/issues/864

LonelyCat124 commented:
I installed GASNet today and am currently working out how to run Regent with it (requires me to save an executable from Regent, so have to get my link flags correct).

LonelyCat124 commented:
Ok, so I ran the 2048x2048 case across 2 nodes today! I submitted it with:
```
bsub -q SKL -n 2 -R "span[ptile=1]" -W 01:00 -e error -o output "blaunch ./nemolite2D -ll:cpu 28 -ll:util 4 -ll:csize 4096 -t1 256 -t2 1024 > out"
```
and it took 144.769485 - which is about the same time (a few seconds faster than the fastest recorded) as the single node runs. I've also been looking into the legion runtime and the default mapper to get to grips better with how that all works, and once I finish goiing through the default mapper I'll look at how some of the customised mappers for other examples work, and go from there. I also tested a naive change (switching tilesize 2 from 1024 to 512) and the runtime doubled, and reducing tilesize1 also increased the runtime - this seems like expected behaviour from previous experience with task sizes.

I also tried weak scaling to 4096x4096 (I guess this is wrong since i double both dimensions i quadruple the work), requesting `-ll:csize 16000` to make sure there was sufficient memory. This took 1030 seconds with 256 and 1024, 644/694 seconds with 256/2048, and 554.646s with 512/2048.

The Intel compiled OpenMP versoin took 612.58s on a single node, while the Regent version took 681s on a single node with 256/2048, and 556s with 512/2048.

At some point I'd like to test running on the GPU as well - my concerns are whether GPU versions of the code can be created yet - I may ask James to install llvm-3.8 (or 6 if that fails) on glados and if I can get an account.

There are also some flags I'd like to tune regarding how many tasks are mapped/readied in advance of execution which could improve performance if there are sufficient util threads I think. I also should profile the code to see how much stuff is running on each node, as I'm not sure its set up great for that right now, based on what I've read of the default mapper.

arporter commented:
@LonelyCat124 When you say "whether GPU versions of the code can be created yet" do you mean with Regent or with PSyclone?

LonelyCat124 commented:
@arporter I mean specifically my version with Regent

LonelyCat124 commented:
Currently working on trying to get a profile for my multinode run - its certainly possible based on my current results that the multinode runs are technically launching on multiple nodes but doing all the work on the primary node at this point.

LonelyCat124 commented:
@arporter @sergisiso  There's an optimization mentioned in the Fortran OpenMP
```
! We have to block here since sshn_t is used in the following loop.
! We could avoid this by altering the following two loop nests to read from
! ssha%data instead of sshn_t%data.
```

I could easily do this in the regent version, and I think it would give some decent benefits as (from the long ago profiling) the update task is currently done in serial. Would it be fair to do this as the OpenMP does not - but the OpenMP version does parallelise this updates normally.

I also realised that I can (and possibly should) use the copy operation in Regent to do this update task - though there are some caveats in the use of copy it seems at the moment. If I can use this these will use the more heavily optimised copy tasks inside regent instead of actually copying data I think.

arporter commented:
That's a good spot (presumably by @sergisiso?). However, I don't think it would be fair to do it in Regent without doing it also in OpenMP. It would be interesting to see the effect in OpenMP and consider whether we could identify this opportunity in PSyclone. (In fact, at present, we don't attempt to reduce thread synchronisation at all - it's on the TODO list.)

On the other hand, using Regent's copy operation is an entirely valid thing to do and the kind of thing PSyclone would implement as a 'built-in'.

LonelyCat124 commented:
I'm happy to add it into the Fortran OpenMP version, as it should be fairly straightforward if I understand the comment correctly - I'll create a corresponding issue.

sergisiso commented:
> That's a good spot (presumably by @sergisiso?).

Nope, a git blame says that comment has been there fore 6 years :)

LonelyCat124 commented:
Definitely have an issue with my GASNet + Legion installation at the moment so the multinode results above are not accurate :)

LonelyCat124 commented:
Ok, so I added the optimisation from the OpenMP code into the Regent code also.
I also tried adding copy's instead of another task.

With the copys (3 runs for each as variation is not that small right now)
| cpu | util | runtime (s) |
|-----|------|-------------|
| 24 | 4 | 176.665394 |
| 24 | 4 | 171.521977 |
| 24 | 4 | 174.495125 |
| 24 | 8 | 180.042847 |
| 24 | 8 | 172.333033 |
| 24 | 8 | 176.814703 |
| 28 | 4 | 177.809224 |
| 28 | 4 | 177.212969 |
| 28 | 4 | 168.515200 |
| 28 | 8 | 237.809098 |
| 28 | 8 | 216.614292 |
| 28 | 8 | 236.150540 |
| 32 | 4 | 174.425761 |
| 32 | 4 | 175.734267 |
| 32 | 4 | 174.555342 |
| 32 | 8 | 180.577154 |
| 32 | 8 | 180.189149 |

With the task
| cpu | util | runtime (s) |
|-----|------|-------------|
| 24 | 4 | 153.670566 |
| 24 | 4 | 153.354698 |
| 24 | 4 | 153.083712 |
| 24 | 8 | 156.463933 |
| 24 | 8 | 157.859750 |
| 24 | 8 | 155.327893 |
| 28 | 4 | 137.130624 |
| 28 | 4 | 135.100195 |
| 28 | 4 | 151.500651 |
| 28 | 8 | 147.391249 |
| 28 | 8 | 156.234955 |
| 28 | 8 | 161.089620 |
| 32 | 4 | 156.202798 |
| 32 | 4 | 157.391080 |
| 32 | 4 | 155.791138 |
| 32 | 8 | 165.358624 |
| 32 | 8 | 160.707995 |

LonelyCat124 commented:
Started implementing the checksum function using metaprogramming (as i want to learn it...is useful for DSLs generally :P )- most of the way there just struggling with unique names for tasks.

LonelyCat124 commented:
Ok, the checksum tasks are implemented (though there are some issues right now, due to rdir but somewhere along my development i introduced a bug I think, gonna try to find the error, assume it was when i redid the partitions, I guess I'll just revert commits one at a time until i find a working version :(

LonelyCat124 commented:
Ok, I reran the OpenMP code with gcc and `-O2` and now the regent code is correct again according to the output, tbut the checksums are not quite the same.

My checksum comparisons:
Fortran:
```
ua checksum =   0.27785892E+05
va checksum =   0.57543179E+05
```

Regent:
```
ua checksum 2.778589e+04
va checksum 6.006844e+04
```

My previous correctness checking tool does the following:
Looks at the output of `go2d_03000_00001.dat` from the two codes, and reads them into python, and then compares the data frames with `print(base_data.equals( new_data))`, which currently prints `True`.

@arporter can I check this is a correct implementation of the and if you have any idea what might be causing the issue? I guess i've done something wrong in updating v_after in the velocity, but not clear how that wouldn't cause issues elsewhere that I should catch:
```
task checksum_task( input_field : region(ispace(int2d), double)) : double
where reads(input_field) do

  var sum : double = 0.0
  for x in input_field do
    sum = sum + abs(input_field[x])
  end
  return sum
end
```

arporter commented:
I'm just about to take my midday break/home-schooling time but as a first thought, I would plot the fields from both versions and check whether they are similar or wildly different (the disadvantage of a checksum is that it provides limited info).

LonelyCat124 commented:
Hi Andy, do you mean the fields output into go2d_03000_...dat?

arporter commented:
> 
> 
> Hi Andy, do you mean the fields output into go2d_03000_...dat?

Yep.

LonelyCat124 commented:
Those files are identical (according to python at least) after 3k steps.

arporter commented:
Hmm, in that case I think the only thing it can be is a difference in the region of the field that is included in the checksum - i.e. the boundary points. In the Fortran version it is only done for the internal points:

    val = array_checksum(field%get_data(), field%data_on_device,           &
                         field%internal%xstart, field%internal%xstop,      &
                         field%internal%ystart, field%internal%ystop)

so possibly there are issues with the `field%internal%{x,y}{start,stop} values not agreeing with the Region domain?

LonelyCat124 commented:
That could be it, having just printed the internal points they go from 2 to 1025, wheras i assume I'm going from 1 to 1027, so I'll fix the checksum there and see.

LonelyCat124 commented:
```
ua checksum 2.778589e+04
va checksum 5.754318e+04
```
That fixed it (at least for 1024x1024).

@arporter I think this is pretty much ready to go now from a "working and performance pov" (other performance things are held up in Regent issues I believe - In terms of clean up which of the following would you think would be worthwhile before opening it up to review:

    1) Remove the commented out fortran code
    2) Remove the python correctness test scripts
    3) Add the regent build command to the Makefile
    4) Update/fix the README.md
    5) Improve the comments explaining what the code does?
    6) Delete dead code.

4 and 6 I definitely need to do, and 5 I would like to do, which of the others would you prefer (slash who should I set to review this?)

arporter commented:
Great! I think we need all of those except for 2)  - assuming the checksums are now output by the Regent version? As for reviewer, it could be me, @sergisiso or @rupertford.

LonelyCat124 commented:
I guess does anyone actively want to review this? If not I'll probably go with sergi if he has time.

arporter commented:
> 
> I guess does anyone actively want to review this? If not I'll probably go with sergi if he has time.

Our working practise (in PSyclone) is to request reviews from multiple people (if appropriate) and then it's a race to see who picks it up :-) (Albeit maybe a 'slow race' where the objective is to come last...)

sergisiso commented:
I am happy to review it since I am interested to have a look at Regent/Legion and have a more broad view of the ECP projects we have. But the PR is quite daunting in size (an with an unfamiliar language) so you will need to be patient with me :)

Regarding 2) if you already have this correctness python scripts I wouldn't remove them.

LonelyCat124 commented:
I finally actually ran a multi-node job! The performance side of it isn't working yet (building Regent with it causes issues on 1 or many nodes right now for me) but I'm working on that, but it works at least!

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/regent/update_values.rg:
```diff
@@ -0,0 +1,68 @@
+                               sea_surface_halos : region(ispace(int2d), uvt_time_field) )
+                     where writes(sea_surface.u_now), reads(grid_region.area_u, grid_region.area_t, grid_region.tmask, sea_surface_halos.t_after) do
+
+--If statement prevents vectorisation at current
```
Same in comment below

LonelyCat124 replied:
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/regent/Makefile:
```diff
@@ -0,0 +1,16 @@
+libgocean2d_io_mod.so : kind_params_mod.o gocean2d_io_mod.o
+	$(F90) kind_params_mod.o gocean2d_io_mod.o -shared -o $@
+
+nemolite2d.exe: 
```
Shouldn't this have all *.rg as target dependencies or be a PHONY target?

LonelyCat124 replied:
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/Makefile:
```diff
@@ -0,0 +1,16 @@
+F90 = gfortran
```
I'm not sure at this point if it would support other compilers (as it relies on libgfortran). I'd rather just add to the comments that only that works for now.

sergisiso replied:
That's a good point. So then, can you write that gfortran is a requirement due to importing libgfortran during the execution in the top-level comment for documentation.

And you will need to remove the F90FLAG that I suggested in the previous review (oops) as this could contain flags for other compilers.

LonelyCat124 replied:
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/Makefile:
```diff
@@ -0,0 +1,16 @@
+F90 = gfortran
+
+%.o : %.f90
+	$(F90) -fPIC -c -o $@ $<
```
I think fPIC is required - i get this error when trying to create the shared object without it:
```
/lustre/scafellpike/local/apps/gcc/utilities/bin/ld: gocean2d_io_mod.o: relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
/lustre/scafellpike/local/apps/gcc/utilities/bin/ld: final link failed: nonrepresentable section on output
/usr/bin/sha1sum: libgocean2d_io_mod.so: No such file or directory
collect2: error: ld returned 1 exit status
```

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/Makefile:
```diff
@@ -0,0 +1,16 @@
+F90 = gfortran
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/Makefile:
```diff
@@ -0,0 +1,16 @@
+regent: nemolite2d.exe
+
+libgocean2d_io_mod.so : kind_params_mod.o gocean2d_io_mod.o
+	$(F90) kind_params_mod.o gocean2d_io_mod.o -shared -o $@
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/README.md:
```diff
@@ -0,0 +1,52 @@
+`ll:csize Z` requests Z MB of CPU DRAM (per node) for the application, without this being set to a sufficient value the program will fail to run, though the default is sufficient for the 258x258 case.
+There are also some specific parameters to tune the application:
+`t1 A` and `t2 B` controls how the system is partitioned into work across the node. We would recommend starting with `A=256` and `B = problemsize/2`.
+
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/README.md:
```diff
@@ -0,0 +1,52 @@
+x-ordinate, y-ordinate, depth, sea-surface height, u, v
+
+where u and v are the x and y components of velocity, respectively.
+
```
Done

sergisiso replied:
I can't see this change. To be more specific I meant adding a bullet point in `benchmarks/nemo/nemolite2d/manual_versions/README.md` at line 44

LonelyCat124 replied:
I forgot to push it, I'll push it this time ^_^

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/algorithm.rg:
```diff
@@ -0,0 +1,381 @@
+import "regent"
```
Fixed I think

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/algorithm.rg:
```diff
@@ -0,0 +1,381 @@
+  if(local_y < 1) then
+    local_y = 1
+  end
+  var partition_space = ispace(int2d, {x = local_x, y = local_y})
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/algorithm.rg:
```diff
@@ -0,0 +1,381 @@
+                         i)
+    end
+    for part in partition_space3 do
+    update_vvel_boundary(update_vvel_bound_partitions[part][0],
```
I think it shouldn't make any difference (since its all asynchronous anyway)

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/calculate_sea_surface_height_t.rg:
```diff
@@ -0,0 +1,42 @@
+
+local c = regentlib.c
+
+--This is the FIRST loop.
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/calculate_velocity_fields.rg:
```diff
@@ -0,0 +1,330 @@
+  return int64_as_double(  ( double_as_int64(x) and not(1LL << 63)) ^ (double_as_int64(y) and (1LL << 63) ) )
+end
+
+--This is the SECOND loop.
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/gocean2d_io_mod.f90:
```diff
@@ -0,0 +1,80 @@
+module gocean2d_io_mod
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/kind_params_mod.f90:
```diff
@@ -0,0 +1,14 @@
+  public
+
+  !> Douple precision kind parameter
+  integer, parameter :: GO_WP = c_double !SELECTED_REAL_KIND(12,307)
```
Is the dl_esm_inf library safe to use for the C versions without this definition btw? Should we not be using interoperable types in that anyway?

sergisiso replied:
dl_esm_inf is very Fortran-centric (like the other infrastructure libraries that PSyclone deals with and we don't have direct control over like LFRic) so it is probably OK as it is but we can consider the change if it is useful. Anyway this is probably a discussion for a separate issue.

But for Regent I see gocean2d_io_mod is an isolated copy that doesn't use the dl_esm_inf infrastructure, so I think it would be OK to replace the types directly there to use the c bindings. Alternatively, you can keep the `kind_params_mod.f90` but making clear in a top comment that this imitate the naming of dl_esm_inf types but it uses C types instead.

LonelyCat124 replied:
Did the latter.

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/initialise_grid_points.rg:
```diff
@@ -0,0 +1,102 @@
+end
+
+--Initialise the sea surface values
+task init_surface_now( sea_surface : region(ispace(int2d), uvt_time_field), grid_region : region(ispace(int2d), grid_fields) ) where
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/model_init.rg:
```diff
@@ -0,0 +1,338 @@
+import "regent"
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/model_init.rg:
```diff
@@ -0,0 +1,338 @@
+  gphi_v : double,
+  xt : double,
+  yt : double
+-- gphi_f  : double  
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/model_write.rg:
```diff
@@ -0,0 +1,37 @@
+local c = regentlib.c
+
+--Writes out the output using c function calls.
+task model_write(step: int, sea_surface : region(ispace(int2d), uvt_time_field), sea_bed_to_mean_sea_level : region(ispace(int2d), uvt_field),
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/namelist:
```diff
@@ -0,0 +1,16 @@
+!-----------------------------------------------------------------------
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/read_config.rg:
```diff
@@ -0,0 +1,56 @@
+config = terralib.types.newstruct("config")
+config.entries:insertall(config_fields_tiles)
+
+--Taken from pennant_common.rg
```
I'm not sure if I can do this while meeting the Apache license rules (since I don't want to add an extra license). I assume I should probably rewrite this code?

sergisiso replied:
Oh that's a tricky question, I think it is small and general enough to be fair use but to be in the safe side it may be better to rewrite it as you say.

LonelyCat124 replied:
It is admittedly difficult to rewrite as I'm not really sure what one might change from this function. @rupertford any idea regarding this? I think the Apache v2 license is used.

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/read_config.rg:
```diff
@@ -0,0 +1,56 @@
+     return quote conf.[field.field] = [field.default_value] end
+   end)]
+
+  --For each of the arguments in the config input, get the value using get_optional_arg and store it in the config structure
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/read_namelist.rg:
```diff
@@ -0,0 +1,73 @@
+ 
+task print_setup_data(output: region(ispace(int1d), setup_type))
+where reads(output) do
+ stdio.printf("%i %i %f %f %i %i %i %i %f %f %f %f\n", output[0].jpiglo, output[0].jpjglo, output[0].dx, output[0].dy, output[0].nit000, output[0].nitend, output[0].record, output[0].jphgr_msh, output[0].dep_const, output[0].rdt, output[0].cbfr, output[0].visc); 
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/test_result.py:
```diff
@@ -0,0 +1,29 @@
+import numpy as np
```
Done. I reduced the amount of output (but left things in to enable comparison in comments, but definitely they could be removed).

LonelyCat124 commented:
I think I did all the reviews except for those I responded to with questions about

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/README.md:
```diff
@@ -0,0 +1,60 @@
+# Regent version of NEMOLite2D #
+
+This implementation of NEMOLite2D uses Regent (http://regent-lang.org/) to handle the dataflow parallemlism.
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/README.md:
```diff
@@ -0,0 +1,60 @@
+CUDA 5.0 or later (for GPUs, yet to be implemented/tested)
+
+## Building ##
+The Makefile picks up the compiler and associated flags from environment
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/algorithm.rg:
```diff
@@ -0,0 +1,374 @@
+                             cbfr,
+                             d2r)
+    end
+-- This is currently commented out as Regent cannot yet index launch two-level projection functors (#845)
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/calculate_velocity_fields.rg:
```diff
@@ -0,0 +1,330 @@
+                                           velocity_full.{u_now,v_now},
+                       sea_bed_to_mean_sea_level.{u,t,v}, sea_surface.{u_now,v_now,t_now},
+                       sea_surface.u_after) do
+--This loop cannot yet vectorize due to copysign implementation (#864) and some remaining if statements (#859)
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/model_write.rg:
```diff
@@ -0,0 +1,41 @@
+        var point : int2d = int2d({x,y})
+        var rtmp1 = 0.5 * (velocity[point + {-1,0}].u_now + velocity[point].u_now) 
+        var rtmp2 = 0.5 * (velocity[point + {0,-1}].v_now + velocity[point].v_now)
+        c.fprintf(f, "%16.6e %16.6e %16.6e %16.6e %16.6e %16.6e\n", grid[point].xt, grid[point].yt, sea_bed_to_mean_sea_level[point].t, sea_surface[point].t_now, rtmp1, rtmp2)
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/read_namelist.rg:
```diff
@@ -0,0 +1,75 @@
+  var setup_data = region(setup_data_is, setup_type)
+  fill(setup_data.{jpiglo, jpjglo, nit000, nitend, record, jphgr_msh}, 0)
+  fill(setup_data.{dx, dy, dep_const, rdt, cbfr, visc}, 0)
+--  init(setup_data)
```
I think this main function can be removed now, its dead code.

LonelyCat124 replied:
I think print_setup_data also is, so I think I'll kill that too

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/test.sh:
```diff
@@ -0,0 +1,20 @@
+
```
Done

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/tmask.py:
```diff
@@ -0,0 +1,18 @@
+import numpy as np
```
Done

LonelyCat124 replied:
Actually missed this. Will be done shortly

LonelyCat124 replied:
Actually I may just remove this file (as it was to test I setup the tmask correctly initially). Woudl that be ok @sergisiso ?

sergisiso replied:
Sure, removing it is OK.

LonelyCat124 replied:
Removed

LonelyCat124 commented on benchmarks/nemo/nemolite2d/manual_versions/regent/test_result.py:
```diff
@@ -0,0 +1,38 @@
+#Takes a single argument which is the timestep to compare, e.g. to compare results after 3000 steps:
```
Done

LonelyCat124 commented:
I think this is all ready for review except for the licence concern I've not touched yet, as I'm not clear how to rewrite it yet.

sergisiso commented:
@LonelyCat124 I marked as resolved most conversations and put the reviewed with actions label again even though no new actions are requested, just two previous comments that are still open.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/regent/README.md:
```diff
@@ -59,3 +59,7 @@ x-ordinate, y-ordinate, depth, sea-surface height, u, v
 
+## License Note##
+Most of this implementation is under the BSD 3-Clause license, unless otherwise
+noted. The Apache-2 License in this folder appliked to one of the functions in
```
I suggest this text: "Most of this implementation is under the BSD 3-Clause license available in the repository root directory, the exception is the 'get_optional_arg' function in read_config.rg that has been copied from https://github.com/StanfordLegion/legion and is distributed with the Apache-2 licence file available in this directory."

And maybe also rename the LICENSE.txt in this directory to LICENSE.apache_v2.txt to avoid confusions.

LonelyCat124 replied:
Done

LonelyCat124 commented:
Should be good to go
